src/utils/wallet.js

```diff
@@ -70,26 +74,68 @@ export function catchTransactionError(e: Object, type: string, tx: Object) {
 }
 
 // handle eth_signTransaction
-export async function signTransaction(trx: Object, wallet: Object): Promise<string> {
+export function signTransaction(trx: Object, wallet: Object): Promise<string> {
   wallet.provider = providers.getDefaultProvider(NETWORK_PROVIDER);
   if (trx && trx.from) {
     delete trx.from;
   }
-  const result = await wallet.sign(trx);
-  return result;
+  return wallet.sign(trx);
 }
 
 // handle eth_sign
-export async function signMessage(message: any, wallet: Object): Promise<string> {
+export function signMessage(message: any, wallet: Object): string {
   wallet.provider = providers.getDefaultProvider(NETWORK_PROVIDER);
   // TODO: this method needs to be replaced when ethers.js is migrated to v4.0
-  const result = ethSign(message, wallet.privateKey);
-  return result;
+  return ethSign(message, wallet.privateKey);
 }
 
 // handle personal_sign
-export async function signPersonalMessage(message: string, wallet: Object): Promise<string> {
+export function signPersonalMessage(message: string, wallet: Object): Promise<string> {
   wallet.provider = providers.getDefaultProvider(NETWORK_PROVIDER);
-  const result = await wallet.signMessage(isHexString(message) ? ethers.utils.arrayify(message) : message);
-  return result;
+  return wallet.signMessage(isHexString(message) ? ethers.utils.arrayify(message) : message);
+}
+
+// we use basic AsyncStorage implementation just to prevent backup being stored in same manner
+export async function getWalletFromStorage(dispatch: Dispatch, appSettings: Object) {
+  let { wallet = {} } = await storage.get('wallet');
+  const walletBackup = await AsyncStorage.getItem(WALLET_STORAGE_BACKUP_KEY);
+  const isWalletEmpty = !wallet || !Object.keys(wallet).length;
+  // wallet timestamp missing causes welcome screen
+  let walletTimestamp = appSettings.wallet;
+  if (isWalletEmpty && walletBackup) {
+    console.log('RESTORING WALLET FROM BACKUP');
+    // restore wallet to storage
+    wallet = JSON.parse(walletBackup);
+    dispatch(saveDbAction('wallet', { wallet }));
+  }
+  // we can only set new timestamp if any wallet is present (existing or backup)
+  if (!walletTimestamp && (!isWalletEmpty || walletBackup)) {
+    walletTimestamp = +(new Date());
+    console.log('SETTING NEW WALLET TIMESTAMP');
+    // only wallet timestamp was missing, let's update it to storage
+    dispatch(saveDbAction('app_settings', { appSettings: { wallet: walletTimestamp } }));
+  }
+  // we check for previous value of `appSettings.wallet` as by this point `walletTimestamp` can be already set
+  if (isWalletEmpty || !appSettings.wallet) {
+    Sentry.captureMessage('Wallet login issue spotted', {
+      extra: {
+        missing: {
+          wallet: isWalletEmpty,
+          walletTimestamp: !appSettings.wallet,
```

Comment:
Maybe we can call these the same? That way it would make more sense when comparing the entry with the code. Like `missing: { isWalletEmpty, walletTimestamp, ... }`.

Reply:
I would call it `walletCreationTimestamp`

Reply:
Updated.